src/rgw/driver/sfs/bucket.cc

```diff
@@ -148,149 +147,96 @@ int SFSBucket::list(
+  const bool want_list_versions =
+      versioning_enabled() ? params.list_versions : false;
+  const bool listing_succeeded =
+      (want_list_versions && list.versions(
+                                 get_bucket_id(), params.prefix, start_with,
+                                 max, results.objs, &results.is_truncated
+                             )) ||
+      list.objects(
           get_bucket_id(), params.prefix, start_with, max, results.objs,
           &results.is_truncated
-      )) {
-    if (!params.delim.empty()) {
-      std::vector<rgw_bucket_dir_entry> new_results;
-      list.roll_up_common_prefixes(
-          params.prefix, params.delim, results.objs, results.common_prefixes,
-          new_results
       );
+  if (!listing_succeeded) {
+    lsfs_dout(dpp, 10) << fmt::format(
+                              "list (prefix:{}, start_after:{}, "
+                              "max:{}) failed.",
+                              params.prefix, start_with, max,
+                              results.objs.size(), results.next_marker,
+                              results.is_truncated
+                          )
+                       << dendl;
+    return -ERR_INTERNAL_ERROR;
+  }
 
-      // Is there actually more after the rolled up common prefix? We
-      // can't tell form the original object query. Ask again for
-      // anything after the prefix.
-      if (!results.common_prefixes.empty()) {
-        std::vector<rgw_bucket_dir_entry> objects_after;
-        std::string query = std::prev(results.common_prefixes.end())->first;
-        query.append(
-            sfs::S3_MAX_OBJECT_NAME_BYTES - query.size(),
-            std::numeric_limits<char>::max()
-        );
-        list.objects(get_bucket_id(), params.prefix, query, 1, objects_after);
-        results.is_truncated = objects_after.size() > 0;
-      }
-      lsfs_dout(dpp, 10) << fmt::format(
-                                "common prefix rollup #objs:{} -> #objs:{}, "
-                                "#prefix:{}, more:{}",
-                                results.objs.size(), new_results.size(),
-                                results.common_prefixes.size(),
-                                results.is_truncated
-                            )
-                         << dendl;
-      results.objs = new_results;
-    }
-    if (results.is_truncated) {
-      if (results.common_prefixes.empty()) {
-        const auto& last = results.objs.back();
-        results.next_marker = rgw_obj_key(last.key.name);
-      } else {
-        const std::string& last =
-            std::prev(results.common_prefixes.end())->first;
-        results.next_marker = rgw_obj_key(last);
-      }
+  if (!params.delim.empty()) {
+    std::vector<rgw_bucket_dir_entry> new_results;
+    list.roll_up_common_prefixes(
+        params.prefix, params.delim, results.objs, results.common_prefixes,
+        new_results
+    );
```

Comment:
From what I can tell, won't this essentially filter the `result.objs` contents? If that's the case, isn't it possible that we will end up with less than `max` entries in the resulting `results.objs = new_results` (later in the function), even if the results have been truncated during the listing performed above?